geth/node/manager.go

```diff
@@ -70,13 +76,91 @@ func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, err
 		return nil, ErrNodeExists
 	}
 
-	ethNode, err := MakeNode(config)
+	m.nodeSyncStarted = make(chan struct{}, 0)
+	m.nodeSyncError = make(chan struct{}, 0)
+	m.nodeSyncCompleted = make(chan struct{}, 0)
+
+	ethNode, err := MakeNode(config, func(leth *les.LightEthereum, stack *node.Node, config *params.NodeConfig) {
+		downlder := leth.Downloader()
+
+		go func() {
+			syncStateWait := time.NewTimer(params.DelayCycleForSyncStart)
+			syncStateWaitMax := time.NewTimer(params.DelayCycleForSyncStartMaxWait)
+
+			defer syncStateWait.Stop()
+			defer syncStateWaitMax.Stop()
+
+			// Initially we want to listen for when node have started synchronizing it's data.
+		nloop:
+			for {
+				select {
+				case <-m.nodeSyncError:
+					// Listen for error to ensure we are not in a case of concurrent runs of this,
+					// as I noticed an issue of double closure of channels, but no way of knowing why.
+					// This provides us a benefit that the go-routine will instantly be shutdown if the channel
+					// was already closed.
+					return
+				case <-syncStateWaitMax.C:
+					close(m.nodeSyncError)
+					return
+				case <-syncStateWait.C:
+					if downlder.Synchronising() {
+						close(m.nodeSyncStarted)
+						break nloop
+					}
+					syncStateWait.Reset(params.DelayCycleForSyncStart)
+				}
+			}
+
+			syncStateWait.Reset(params.DelayCycleForSyncCompleted)
+			syncStateWaitMax.Reset(params.DelayCycleForSyncCompletedMaxWait)
+
+			// Once we have validating, that things have properly started then, we then track
+			// progress for when a finished signal is called.
+		cloop:
+			for {
+				select {
+				case <-m.nodeSyncError:
+					// Listen for error to ensure we are not in a case of concurrent runs of this,
+					// as I noticed an issue of double closure of channels, but no way of knowing why.
+					// This provides us a benefit that the go-routine will instantly be shutdown if the channel
+					// was already closed.
+					return
+				case <-syncStateWaitMax.C:
+					close(m.nodeSyncError)
```

Comment:
`downloader.Synchronising()` returns `true/false` as to when the sync has started, we at the first poll for that delay if sink started, since we can't know till this is true. Using this strategy, we can equally close the `m.nodeSyncError` channel to signal a failure to either start sync or complete sync in provided allowed duration.

If we want more context, then one more channel is needed to different when a node does not start syncing and when a node fails to complete sync within set max duration.

Reply:
I guess we don't need that information. We can encapsulate sync to just one step as you did even though it's two steps. Maybe we can just add logs?